File: dc/internal/model/agent_guard.go
package model

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// AgentBehaviorEvent is the immutable, query-oriented projection of one
// aegis.agent_behavior.v1 runtime event. Raw RuntimeEvent persistence remains
// independent so Kafka replay can retry either write safely.
type AgentBehaviorEvent struct {
	ID                     uuid.UUID       `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	RawEventID             string          `gorm:"column:raw_event_id;uniqueIndex;not null" json:"raw_event_id"`
	HostID                 uuid.UUID       `gorm:"type:uuid;index;not null" json:"host_id"`
	HostBootID             string          `gorm:"column:host_boot_id;not null" json:"host_boot_id"`
	AgentSequence          int64           `gorm:"column:agent_sequence;not null" json:"agent_sequence"`
	InstanceID             *uuid.UUID      `gorm:"type:uuid" json:"instance_id,omitempty"`
	SessionID              *uuid.UUID      `gorm:"type:uuid" json:"session_id,omitempty"`
	ExecutionUnitID        *uuid.UUID      `gorm:"type:uuid" json:"execution_unit_id,omitempty"`
	PolicyID               *uuid.UUID      `gorm:"type:uuid" json:"policy_id,omitempty"`
	PolicyVersion          *int64          `json:"policy_version,omitempty"`
	RuleID                 string          `json:"rule_id,omitempty"`
	SchemaVersion          string          `gorm:"column:schema_version;not null" json:"schema_version"`
	CorrelationID          string          `json:"correlation_id,omitempty"`
	ParentEventID          string          `json:"parent_event_id,omitempty"`
	AgentType              string          `json:"agent_type,omitempty"`
	ProfileKey             string          `json:"profile_key,omitempty"`
	ProfileVersion         *int64          `json:"profile_version,omitempty"`
	Category               string          `gorm:"not null" json:"category"`
	Operation              string          `gorm:"not null" json:"operation"`
	Outcome                string          `gorm:"not null" json:"outcome"`
	Errno                  *int            `json:"errno,omitempty"`
	Decision               string          `gorm:"not null" json:"decision"`
	Severity               string          `gorm:"not null" json:"severity"`
	PID                    int             `gorm:"column:pid" json:"pid,omitempty"`
	PPID                   int             `gorm:"column:ppid" json:"ppid,omitempty"`
	ProcessStartTicks      string          `gorm:"column:process_start_ticks" json:"process_start_ticks,omitempty"`
	ProcessName            string          `json:"process_name,omitempty"`
	ProcessExe             string          `json:"process_exe,omitempty"`
	CommandArgv            json.RawMessage `gorm:"column:command_argv;type:jsonb;not null" json:"command_argv"`
	CommandCWD             string          `json:"command_cwd,omitempty"`
	CommandVisibility      string          `gorm:"not null" json:"command_visibility"`
	ProcessChain           json.RawMessage `gorm:"type:jsonb;not null" json:"process_chain"`
	ResourceType           string          `json:"resource_type,omitempty"`
	ResourceIdentity       string          `json:"resource_identity,omitempty"`
	ResourceIdentityHash   string          `json:"resource_identity_hash,omitempty"`
	ResourceClassification string          `json:"resource_classification,omitempty"`
	Resource               json.RawMessage `gorm:"type:jsonb;not null" json:"resource"`
	Isolation              json.RawMessage `gorm:"type:jsonb;not null" json:"isolation"`
	Collection             json.RawMessage `gorm:"type:jsonb;not null" json:"collection"`
	Evidence               json.RawMessage `gorm:"type:jsonb;not null" json:"evidence"`
	OccurredAt             time.Time       `gorm:"not null" json:"occurred_at"`
	OccurredMonotonicNS    *int64          `gorm:"column:occurred_monotonic_ns" json:"occurred_monotonic_ns,omitempty"`
	ReceivedAt             time.Time       `gorm:"not null" json:"received_at"`
	CreatedAt              time.Time       `json:"created_at"`
	AggregatedCount        int64           `gorm:"-" json:"-"`
	LostEventsSinceLast    int64           `gorm:"-" json:"-"`
	HasTruncatedFields     bool            `gorm:"-" json:"-"`
	Completeness           json.RawMessage `gorm:"-" json:"-"`
}

func (AgentBehaviorEvent) TableName() string {
	return "agent_behavior_events"
}

type AgentRuntimeInstance struct {
	ID                   uuid.UUID       `gorm:"type:uuid;primaryKey" json:"id"`
	HostID               uuid.UUID       `gorm:"type:uuid;not null" json:"host_id"`
	AssetID              *uuid.UUID      `gorm:"type:uuid" json:"asset_id,omitempty"`
	ProfileKey           string          `gorm:"not null" json:"profile_key"`
	ProfileVersion       int64           `gorm:"not null" json:"profile_version"`
	AgentType            string          `gorm:"not null" json:"agent_type"`
	DisplayName          string          `json:"display_name,omitempty"`
	ControllerPID        int             `gorm:"column:controller_pid;not null" json:"controller_pid"`
	ControllerStartTicks string          `gorm:"column:controller_start_ticks;not null" json:"controller_start_ticks"`
	ControllerExe        string          `json:"controller_exe,omitempty"`
	RunUID               *int            `gorm:"column:run_uid" json:"run_uid,omitempty"`
	DetectionConfidence  string          `gorm:"not null" json:"detection_confidence"`
	Status               string          `gorm:"not null" json:"status"`
	CoverageLevel        string          `gorm:"not null" json:"coverage_level"`
	CoverageReasons      json.RawMessage `gorm:"type:jsonb;not null" json:"coverage_reasons"`
	Metadata             json.RawMessage `gorm:"type:jsonb;not null" json:"metadata"`
	FirstSeenAt          time.Time       `gorm:"not null" json:"first_seen_at"`
	LastSeenAt           time.Time       `gorm:"not null" json:"last_seen_at"`
	StoppedAt            *time.Time      `json:"stopped_at,omitempty"`
	CreatedAt            time.Time       `json:"created_at"`
	UpdatedAt            time.Time       `json:"updated_at"`
}

func (AgentRuntimeInstance) TableName() string {
	return "agent_runtime_instances"
}

type AgentExecutionUnit struct {
	ID                uuid.UUID       `gorm:"type:uuid;primaryKey" json:"id"`
	HostID            uuid.UUID       `gorm:"type:uuid;not null" json:"host_id"`
	InstanceID        uuid.UUID       `gorm:"type:uuid;not null" json:"instance_id"`
	UnitType          string          `gorm:"not null" json:"unit_type"`
	Fingerprint       string          `gorm:"not null" json:"fingerprint"`
	RootPID           *int            `gorm:"column:root_pid" json:"root_pid,omitempty"`
	RootStartTicks    *string         `gorm:"column:root_start_ticks" json:"root_start_ticks,omitempty"`
	CgroupPath        string          `json:"cgroup_path,omitempty"`
	ContainerID       string          `json:"container_id,omitempty"`
	ContainerRuntime  string          `json:"container_runtime,omitempty"`
	RemoteBackend     string          `json:"remote_backend,omitempty"`
	RemoteExecutionID string          `json:"remote_execution_id,omitempty"`
	RemoteHostRef     string          `json:"remote_host_ref,omitempty"`
	CoverageLevel     string          `gorm:"not null" json:"coverage_level"`
	CoverageReasons   json.RawMessage `gorm:"type:jsonb;not null" json:"coverage_reasons"`
	IsolationBaseline json.RawMessage `gorm:"type:jsonb;not null" json:"isolation_baseline"`
	IsolationActual   json.RawMessage `gorm:"type:jsonb;not null" json:"isolation_actual"`
	IsolationDiff     json.RawMessage `gorm:"type:jsonb;not null" json:"isolation_diff"`
	Status            string          `gorm:"not null" json:"status"`
	FirstSeenAt       time.Time       `gorm:"not null" json:"first_seen_at"`
	LastSeenAt        time.Time       `gorm:"not null" json:"last_seen_at"`
	StoppedAt         *time.Time      `json:"stopped_at,omitempty"`
	CreatedAt         time.Time       `json:"created_at"`
	UpdatedAt         time.Time       `json:"updated_at"`
}

func (AgentExecutionUnit) TableName() string {
	return "agent_execution_units"
}

type AgentBehaviorSession struct {
	ID                   uuid.UUID       `gorm:"type:uuid;primaryKey" json:"id"`
	HostID               uuid.UUID       `gorm:"type:uuid;not null" json:"host_id"`
	InstanceID           uuid.UUID       `gorm:"type:uuid;not null" json:"instance_id"`
	ExecutionUnitID      *uuid.UUID      `gorm:"type:uuid" json:"execution_unit_id,omitempty"`
	ExternalSessionID    *string         `json:"external_session_id,omitempty"`
	Source               string          `gorm:"not null" json:"source"`
	Confidence           string          `gorm:"not null" json:"confidence"`
	CorrelationTokenHash *string         `json:"correlation_token_hash,omitempty"`
	Status               string          `gorm:"not null" json:"status"`
	BehaviorCount        int64           `gorm:"not null" json:"behavior_count"`
	FindingCount         int64           `gorm:"not null" json:"finding_count"`
	Completeness         json.RawMessage `gorm:"type:jsonb;not null" json:"completeness"`
	StartedAt            time.Time       `gorm:"not null" json:"started_at"`
	LastSeenAt           time.Time       `gorm:"not null" json:"last_seen_at"`
	EndedAt              *time.Time      `json:"ended_at,omitempty"`
	CreatedAt            time.Time       `json:"created_at"`
	UpdatedAt            time.Time       `json:"updated_at"`
}

func (AgentBehaviorSession) TableName() string {
	return "agent_behavior_sessions"
}

type AgentGuardStateProjection struct {
	EventType string
	ObjectID  uuid.UUID
	Instance  *AgentRuntimeInstance
	Unit      *AgentExecutionUnit
	Session   *AgentBehaviorSession
	Delivery  *AgentGuardDeliveryStatus
	Action    *AgentGuardAction
}

type AgentGuardDeliveryStatus struct {
	HostID         uuid.UUID
	BundleVersion  int64
	BundleDigest   string
	Status         string
	ErrorCode      string
	LastReportedAt time.Time
}

type AgentGuardAction struct {
	ID                   uuid.UUID       `gorm:"type:uuid;primaryKey" json:"id"`
	CommandID            string          `gorm:"column:command_id;type:varchar(100);uniqueIndex" json:"command_id,omitempty"`
	TriggerFindingID     *uuid.UUID      `gorm:"type:uuid;index" json:"trigger_finding_id,omitempty"`
	HostID               uuid.UUID       `gorm:"type:uuid;not null;index" json:"host_id"`
	InstanceID           *uuid.UUID      `gorm:"type:uuid;index" json:"instance_id,omitempty"`
	ExecutionUnitID      *uuid.UUID      `gorm:"type:uuid;index" json:"execution_unit_id,omitempty"`
	Action               string          `gorm:"type:varchar(40);not null" json:"action"`
	Source               string          `gorm:"type:varchar(32);not null" json:"source"`
	Status               string          `gorm:"type:varchar(32);not null;index" json:"status"`
	Reason               string          `gorm:"type:text;not null" json:"reason"`
	RequestedBy          string          `gorm:"type:varchar(100)" json:"requested_by,omitempty"`
	HoldRequested        bool            `gorm:"not null" json:"hold_requested"`
	FreezeTimeoutSeconds *int            `json:"freeze_timeout_seconds,omitempty"`
	Result               json.RawMessage `gorm:"type:jsonb;not null" json:"result"`
	ErrorCode            string          `gorm:"type:varchar(100)" json:"error_code,omitempty"`
	ErrorMessage         string          `gorm:"type:text" json:"error_message,omitempty"`
	RequestedAt          time.Time       `gorm:"not null" json:"requested_at"`
	DispatchedAt         *time.Time      `json:"dispatched_at,omitempty"`
	CompletedAt          *time.Time      `json:"completed_at,omitempty"`
	ExpiresAt            *time.Time      `json:"expires_at,omitempty"`
	CreatedAt            time.Time       `json:"created_at"`
	UpdatedAt            time.Time       `json:"updated_at"`
}

func (AgentGuardAction) TableName() string {
	return "agent_guard_actions"
}

type AgentGuardPolicy struct {
	ID                   uuid.UUID       `gorm:"type:uuid;primaryKey"`
	Version              int64           `gorm:"not null"`
	Status               string          `gorm:"type:varchar(32);not null"`
	CorrelationRules     json.RawMessage `gorm:"type:jsonb;not null"`
	AtomicRules          json.RawMessage `gorm:"type:jsonb;not null"`
	EscapeRules          json.RawMessage `gorm:"type:jsonb;not null"`
	FreezeTimeoutSeconds int             `gorm:"not null"`
	PublishedAt          *time.Time
}

func (AgentGuardPolicy) TableName() string {
	return "agent_guard_policies"
}

type AgentSecurityFinding struct {
	ID                  uuid.UUID       `gorm:"type:uuid;primaryKey" json:"id"`
	FindingKey          string          `gorm:"type:varchar(255);not null;uniqueIndex" json:"finding_key"`
	HostID              uuid.UUID       `gorm:"type:uuid;not null;index" json:"host_id"`
	InstanceID          *uuid.UUID      `gorm:"type:uuid;index" json:"instance_id,omitempty"`
	SessionID           *uuid.UUID      `gorm:"type:uuid;index" json:"session_id,omitempty"`
	ExecutionUnitID     *uuid.UUID      `gorm:"type:uuid;index" json:"execution_unit_id,omitempty"`
	PolicyID            *uuid.UUID      `gorm:"type:uuid;index" json:"policy_id,omitempty"`
	PolicyVersion       *int64          `json:"policy_version,omitempty"`
	Title               string          `gorm:"type:varchar(500);not null" json:"title"`
	Severity            string          `gorm:"type:varchar(20);not null;index" json:"severity"`
	Verdict             string          `gorm:"type:varchar(24);not null" json:"verdict"`
	Confidence          float64         `gorm:"type:numeric(5,4);not null" json:"confidence"`
	Status              string          `gorm:"type:varchar(24);not null;index" json:"status"`
	DecisionSources     json.RawMessage `gorm:"type:jsonb;not null" json:"decision_sources"`
	RuleHits            json.RawMessage `gorm:"type:jsonb;not null" json:"rule_hits"`
	EvidenceEventIDs    json.RawMessage `gorm:"type:jsonb;not null" json:"evidence_event_ids"`
	EvidenceGraph       json.RawMessage `gorm:"type:jsonb;not null" json:"evidence_graph"`
	AttackStages        json.RawMessage `gorm:"type:jsonb;not null" json:"attack_stages"`
	Summary             string          `gorm:"type:text" json:"summary,omitempty"`
	RecommendedAction   string          `gorm:"type:varchar(64)" json:"recommended_action,omitempty"`
	FirstObservedAt     time.Time       `gorm:"not null" json:"first_observed_at"`
	LastObservedAt      time.Time       `gorm:"not null;index" json:"last_observed_at"`
	CreatedAt           time.Time       `json:"created_at"`
	UpdatedAt           time.Time       `json:"updated_at"`
	EvidenceSourceTable string          `gorm:"-" json:"-"`
}

func (AgentSecurityFinding) TableName() string {
	return "agent_security_findings"
}

File: frontend/src/types/agentGuard.ts
export type AgentGuardCoverage =
  | 'full_enforcement'
  | 'behavior_monitor_escape_enforce'
  | 'monitor_only'
  | 'no_isolation'
  | 'remote_unobservable'
  | 'unsupported'
  | 'unsupported_profile'
  | 'degraded'
  | 'unknown'

export type AgentGuardMode = 'behavior' | 'escape'
export type AgentGuardDetailTab = 'panorama' | 'analysis'

export type AgentConfigFindingSeverity = 'critical' | 'high' | 'medium' | 'low' | string

export interface AgentConfigFinding {
  rule_id: string
  severity: AgentConfigFindingSeverity
  field_path: string
  value?: string
  title: string
  reason: string
  remediation: string
}

export interface AgentConfigFile {
  path: string
  format: string
  status: string
  size?: number
  mode?: string
  modified_at?: string
  sha256?: string
  content?: string
  error?: string
  findings: AgentConfigFinding[]
}

export interface AgentConfigHook {
  file_path: string
  field_path: string
  event: string
  command: string
  executor?: string
  findings: AgentConfigFinding[]
}

export interface AgentConfigAgent {
  agent_type: string
  display_name: string
  files: AgentConfigFile[]
  hooks: AgentConfigHook[]
  finding_count: number
}

export interface AgentConfigScanError {
  stage: string
  message: string
}

export interface AgentConfigScanResult {
  host_id: string
  hostname?: string
  scanned_at: string
  agents: AgentConfigAgent[]
  errors?: AgentConfigScanError[]
  finding_count: number
}

export type AgentRuntimeStatus = 'running' | 'stale' | 'stopped' | 'unknown'

export type AgentGuardRuntimeDispatchStatus =
  | 'not_dispatched'
  | 'pending'
  | 'pending_reconnect'
  | 'dispatched'
  | 'failed'
  | string

export interface AgentGuardHookInjection {
  agent_type: string
  enabled: boolean
  behavior_enabled?: boolean
  escape_enabled?: boolean
  status: string
  error_code?: string
  updated_at?: string
}

export interface AgentGuardRuntimeSettings {
  schema: string
  version: number
  host_id: string
  tool_adapter_enabled: boolean
  session_hook_enabled: boolean
  behavior_policy_enabled: boolean
  escape_policy_enabled: boolean
  injections: AgentGuardHookInjection[]
  dispatch_status: AgentGuardRuntimeDispatchStatus
  dispatch_error_code?: string
  updated_at?: string
}

export type ExecutionUnitType =
  | 'local_process_tree'
  | 'linux_namespace'
  | 'oci_container'
  | 'remote_sandbox'
  | 'whole_process_container'
  | string

export interface PageResult<T> {
  items: T[]
  total: number
}

export interface AgentGuardOverview {
  agent_assets?: number
  running_instances?: number
  monitored_instances?: number
  high_risk_findings?: number
  successful_blocks?: number
  escape_attempts?: number
  frozen_units?: number
  execution_units?: number
  coverage?: Partial<Record<AgentGuardCoverage, number>>
  updated_at?: string
  stale?: boolean
}

export interface AgentGuardAgentSummary {
  agent_scope_key: string
  asset_id?: string
  host: {
    id: string
    hostname: string
    ip: string
  }
  agent_type: string
  display_name: string
  profile_key?: string
  running_instance_count: number
  /** Number of statically collected Claude/Codex sessions for this host/type. */
  session_count?: number
  controller_pids: number[]
  runtime_status: AgentRuntimeStatus
  asset_status?: 'running' | 'stopped' | string
  asset_collected_at?: string
  isolation_types: ExecutionUnitType[]
  coverage_level: AgentGuardCoverage
  coverage_reasons: string[]
  high_risk_finding_count: number
  escape_finding_count: number
  action_status?: string
  last_seen_at?: string
}

export interface AgentRuntimeInstance {
  id: string
  host_id: string
  asset_id?: string
  agent_type: string
  display_name?: string
  profile_key?: string
  profile_version?: number
  controller_pid: number
  controller_start_ticks: string
  status: AgentRuntimeStatus
  coverage_level: AgentGuardCoverage
  coverage_reasons: string[]
  high_risk_finding_count?: number
  last_seen_at?: string
}

export interface AgentBehaviorSession {
  id: string
  host_id: string
  instance_id: string
  execution_unit_id?: string
  external_session_id?: string
  source: string
  confidence: string
  status: string
  behavior_count?: number
  finding_count?: number
  started_at: string
  last_seen_at: string
  permission?: {
    agent_type?: string
    backend?: string
    boundary?: 'enforced' | 'none' | 'no_isolation' | 'remote_unobservable' | string
    class?: 'full_access' | 'restricted' | 'unknown' | string
    permission_mode?: string
    sandbox_mode?: string
    approval_policy?: string
    approval_status?: string
    cwd?: string
    workspace_roots?: string[]
    temp_roots?: string[]
    network_access?: boolean
    sandbox_enabled?: boolean
    workspace_access?: 'none' | 'ro' | 'rw' | string
    allowed_domains?: string[]
    denied_domains?: string[]
    elevated?: boolean
    approval_required?: boolean
    safe_write_root?: string
    remote_execution_id?: string
    complete?: boolean
  }
}

export interface AgentExecutionUnit {
  id: string
  host_id?: string
  instance_id: string
  unit_type: ExecutionUnitType
  fingerprint?: string
  root_pid?: number
  process_count?: number
  root_start_ticks?: string
  cgroup_id?: string
  cgroup_path?: string
  container_id?: string
  container_runtime?: string
  remote_backend?: string
  remote_execution_id?: string
  coverage_level: AgentGuardCoverage
  coverage_reasons: string[]
  status: string
  isolation_baseline: Record<string, unknown>
  isolation_actual: Record<string, unknown>
  isolation_diff: Record<string, unknown>
  first_seen_at: string
  last_seen_at: string
}

export type AgentGuardActionName =
  | 'freeze_execution_unit'
  | 'resume_execution_unit'
  | 'hold_execution_unit'
  | 'kill_execution_unit'
  | 'kill_agent_instance'
  | 'auto_resume'

export type AgentGuardActionStatus =
  | 'pending'
  | 'dispatching'
  | 'running'
  | 'success'
  | 'failed'
  | 'expired'
  | 'cancelled'
  | string

export interface AgentGuardActionRequest {
  reason: string
  hold?: boolean
}

export interface AgentGuardActionAccepted {
  action_id: string
  command_id: string
  status: AgentGuardActionStatus
}

export interface AgentGuardAction {
  id: string
  command_id?: string
  host_id?: string
  instance_id?: string
  execution_unit_id?: string
  action: AgentGuardActionName
  source?: 'local_policy' | 'correlation_policy' | 'manual' | 'timeout' | 'system' | string
  status: AgentGuardActionStatus
  reason: string
  requested_by?: string
  hold_requested?: boolean
  freeze_timeout_seconds?: number
  result?: Record<string, unknown>
  error_code?: string
  error_message?: string
  requested_at?: string
  dispatched_at?: string
  completed_at?: string
  expires_at?: string
}

export interface PanoramaTreeNode {
  id: string
  parent_id?: string
  node_type: string
  label: string
  severity?: 'info' | 'low' | 'medium' | 'high' | 'critical'
  has_children?: boolean
  child_count?: number
  occurred_at?: string
  pid?: number
  ppid?: number
  process_start_ticks?: string
  process_status?: 'running' | 'stopped' | 'unknown' | string
  cmdline?: string
  external_session_id?: string
  session_source?: 'agent_official' | 'adapter_hook' | 'aegis_wrapper' | 'activity_window' | 'execution_unit' | string
  session_confidence?: 'confirmed' | 'probable' | 'inferred' | string
	tool_name?: string
	tool_call_id?: string
	turn_id?: string
	command?: string
	tool_input?: unknown
	tool_response?: unknown
	correlation_status?: 'matched' | 'unmatched' | string
	correlation_method?: string
  event_id?: string
  object_id?: string
  execution_unit_id?: string
  collection?: {
    visibility?: 'complete' | 'partial' | 'unobservable'
    truncated_fields?: string[]
    lost_events_since_last?: number
    limitations?: Array<'tool_semantics_unobservable' | 'remote_unobservable' | string>
  }
  trust?: {
    tool_semantics?: 'trusted' | 'tool_semantics_unobservable'
    source?: 'agent_official' | 'adapter_hook' | 'aegis_wrapper'
    proof_verified?: boolean
    remote_visibility?: 'trusted_sensor' | 'remote_unobservable'
    correlation?: 'matched' | 'unmatched'
  }
  children?: PanoramaTreeNode[]
}

export interface AgentPanoramaResponse {
  root?: PanoramaTreeNode | null
  items?: PanoramaTreeNode[]
  nodes?: PanoramaTreeNode[]
  total?: number
  next_cursor?: string
}

export interface AgentSecurityFindingSummary {
  id: string
  asset_id?: string
  agent_scope_key?: string
  instance_id?: string
  session_id?: string
  agent_type?: string
  agent_display_name?: string
  host?: {
    id?: string
    hostname?: string
    ip?: string
  }
  title: string
  severity: 'info' | 'low' | 'medium' | 'high' | 'critical'
  verdict?: 'benign' | 'suspicious' | 'malicious' | 'inconclusive'
  confidence?: number
  decision_sources?: Array<'rule' | 'ai' | 'combined'>
  rule_hits?: Array<{
    rule_key?: string
    rule_version?: number
    rule_name?: string
    severity?: string
    match_kind?: string
    event_id?: string
    event_ids?: string[]
    evidence_event_ids?: string[]
  }>
  matched_rules?: AgentSecurityFindingRuleDetail[]
  evidence_event_ids?: string[]
  evidence_event_count?: number
  evidence_graph?: {
    nodes?: Array<Record<string, unknown>>
    edges?: Array<Record<string, unknown>>
  }
  evidence_completeness?: {
    visibility?: 'complete' | 'partial' | 'unobservable'
    reasons?: string[]
    lost_events?: number
    truncated_fields?: string[]
  }
  escape_chain?: {
    hook_event_ids?: string[]
    hook_events?: Array<{
      event_id: string
      event_type?: string
      tool_name?: string
      command?: string
      command_line?: string
      pid?: number
      ppid?: number
      process_start_ticks?: string
      process_name?: string
      process_exe?: string
      cwd?: string
      outcome?: string
      decision?: string
      target?: string
    }>
    process_evidence?: Array<Record<string, unknown>>
    execution_evidence?: Array<Record<string, unknown>>
    permission?: {
      agent_type?: string
      backend?: string
      boundary?: string
      class?: 'full_access' | 'restricted' | 'unknown' | string
      permission_mode?: string
      sandbox_mode?: string
      approval_policy?: string
      approval_status?: string
      cwd?: string
      workspace_roots?: string[]
      temp_roots?: string[]
      network_access?: boolean
      sandbox_enabled?: boolean
      workspace_access?: string
      allowed_domains?: string[]
      denied_domains?: string[]
      elevated?: boolean
      approval_required?: boolean
      safe_write_root?: string
      remote_execution_id?: string
      complete?: boolean
    }
    classification?: 'policy_violation_attempt' | 'confirmed_escape' | 'authorized_boundary_expansion' | 'not_applicable' | 'evidence_insufficient' | string
    gaps?: string[]
  }
  attack_stages?: Array<string | Record<string, unknown>>
  summary?: string
  counter_evidence?: string[]
  uncertainties?: string[]
  recommended_action?: string
  analysis_status?: string
  status: 'open' | 'investigating' | 'contained' | 'resolved' | 'dismissed' | string
  last_observed_at?: string
}

export interface AgentSecurityFindingRuleDetail {
  rule_key: string
  rule_version?: number
  name: string
  severity?: string
  match_kind?: string
  event_ids: string[]
  process_tree?: AgentSecurityFindingProcessNode[]
  tool_calls?: AgentSecurityFindingToolCall[]
}

export interface AgentSecurityFindingToolCall {
  event_id: string
  tool_name: string
  tool_call_id?: string
  turn_id?: string
  command?: string
  tool_input?: unknown
  tool_response?: unknown
  outcome?: string
  occurred_at?: string
  pid?: number
  ppid?: number
  process_start_ticks?: string
  command_line?: string
  correlation_status?: 'matched' | 'unmatched' | string
  correlation_method?: string
}

export interface AgentSecurityFindingProcessNode {
  id: string
  parent_id?: string
  pid: number
  ppid: number
  process_start_ticks?: string
  process_name?: string
  process_exe?: string
  cmdline?: string
  command_cwd?: string
  command_visibility?: string
  process_status?: string
  first_seen_at?: string
  last_seen_at?: string
  event_count?: number
  matched?: boolean
  matched_event_ids?: string[]
  children?: AgentSecurityFindingProcessNode[]
}

export interface AgentBehaviorIndex {
  id: string
  event_id?: string
  asset_id?: string
  agent_scope_key?: string
  instance_id?: string
  agent_type?: string
  agent_display_name?: string
  host?: {
    id?: string
    hostname?: string
    ip?: string
  }
  category?: string
  operation?: string
  outcome?: string
  pid?: number
  ppid?: number
  process_start_ticks?: string
  process_name?: string
  process_exe?: string
  command_argv?: string[]
  command_cwd?: string
  command_visibility?: string
  actor?: Record<string, unknown>
  resource?: Record<string, unknown>
  collection?: Record<string, unknown>
  occurred_at?: string
}

export interface BuiltinAgentBehaviorRuleSummary {
  id?: string
  rule_key: string
  rule_version: number
  name: string
  description?: string
  source?: string
  categories?: string[]
  default_enabled?: boolean
  enabled?: boolean
  engine?: string
  default_severity?: string
  severity?: string
  default_action?: string
  action?: string
  recommended_action?: string
  parameters_schema?: Record<string, unknown>
  default_parameters?: Record<string, unknown>
  required_evidence?: string[]
  allow_conditions?: string[]
  mitre?: string[]
  immutable?: boolean
  hits_24h?: number
  findings_24h?: number
  digest?: string
}

export interface BuiltinAgentEscapeRuleSummary {
  rule_key: string
  rule_version: number
  name: string
  description?: string
  hook_points?: string[]
  required_evidence?: string[]
  agent_types?: string[]
  backends?: string[]
  boundary_semantics?: string[]
  default_enabled?: boolean
  enabled?: boolean
  default_severity?: string
  severity?: string
  default_action?: string
  action?: string
  source?: string
  immutable?: boolean
  digest?: string
}

export interface AgentSecurityAnalysisRun {
  id: string
  finding_id: string
  attempt: number
  status: 'pending' | 'running' | 'completed' | 'failed' | 'invalid_output' | 'inconclusive' | string
  provider?: string
  model?: string
  prompt_version: string
  input_digest: string
  evidence_event_ids: string[]
  verdict?: 'benign' | 'suspicious' | 'malicious' | 'inconclusive'
  attack_probability?: number
  confidence?: number
  output?: {
    summary?: string
    intent_hypotheses?: Array<Record<string, unknown>>
    attack_chain?: Array<Record<string, unknown>>
    counter_evidence?: string[]
    uncertainties?: string[]
    recommended_action?: string
  }
  error_code?: string
  started_at?: string
  completed_at?: string
}

export interface AgentGuardListFilters {
  host_id: string
  agent_types: string[]
  runtime_status: string
  coverage: string
  isolation_type: string
  keyword: string
}

export interface AgentGuardAgentQuery {
  host_ids?: string[]
  agent_types?: string[]
  runtime_status?: string
  coverage?: string
  isolation_type?: string
  keyword?: string
  page: number
  page_size: number
}

export interface AgentGuardInstanceQuery {
  host_id?: string
  asset_ids?: string[]
  agent_scope_key?: string
  agent_types?: string[]
  instance_ids?: string[]
  status?: string
  coverage?: string
  page?: number
  page_size?: number
}

export interface AgentGuardPanoramaQuery {
  asset_id?: string
  agent_scope_key?: string
  instance_ids?: string[]
  session_id?: string
  page?: number
  page_size?: number
  cursor?: string
}

export interface AgentGuardExecutionUnitQuery {
  host_id?: string
  instance_id?: string
  unit_type?: string
  coverage?: string
  status?: string
  page?: number
  page_size?: number
}

export interface AgentGuardFindingQuery {
	host_id?: string
	asset_id?: string
	agent_scope_key?: string
	instance_id?: string
	session_id?: string
	finding_domain?: 'tool' | 'escape'
	severity?: string
  status?: string
  page: number
  page_size: number
}

export interface AgentGuardRuleQuery {
  enabled?: boolean
  keyword?: string
  page?: number
  page_size?: number
}

export interface AgentGuardPolicyTargets {
  host_ids: string[]
  host_group_ids: string[]
  agent_types: string[]
  profile_keys?: string[]
}

export interface AgentGuardPolicyDraftRequest {
  policy_key: string
  name: string
  description?: string
  priority: number
  targets: AgentGuardPolicyTargets
  collection: {
    categories: string[]
    tool_adapter_enabled?: boolean
    command_argv: 'redacted' | 'disabled'
    file_content: 'disabled'
    network_content: 'disabled'
    aggregation: Record<string, number>
  }
  builtin_rule_overrides: Array<Record<string, unknown>>
  atomic_rules: Array<Record<string, unknown>>
  correlation_rules: Array<Record<string, unknown>>
  analysis: {
    enabled: boolean
    trigger_severities: string[]
    ai_only_action_ceiling: 'audit' | 'alert'
    evidence_window_seconds: number
  }
  escape_rules: Array<Record<string, unknown>>
  freeze_timeout_seconds: number
}

export interface AgentGuardPolicy {
  id: string
  policy_key: string
  version: number
  name: string
  description?: string
  status: 'draft' | 'published' | 'superseded' | 'disabled'
  priority: number
  digest?: string
  published_by?: string
  published_at?: string
  updated_at?: string
}

export interface AgentGuardPolicyValidation {
  valid: boolean
  digest?: string
  errors: Array<{ field: string; code: string; message: string }>
  warnings: Array<{ field: string; code: string; message: string }>
}

export interface AgentGuardPolicyMutationResult {
  policy: AgentGuardPolicy
  validation: AgentGuardPolicyValidation
}

export interface AgentGuardPolicyDelivery {
  id: string
  host_id: string
  bundle_version: number
  bundle_digest: string
  status: 'pending' | 'dispatching' | 'received' | 'applied' | 'degraded' | 'failed' | string
  coverage_level?: AgentGuardCoverage
  error_code?: string
  error_message?: string
  generated_at: string
  applied_at?: string
}

export interface AgentGuardPolicyPublishResult {
  policy: AgentGuardPolicy
  deliveries: AgentGuardPolicyDelivery[]
}
